Jint/Runtime/Debugger/DebugScope.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using Jint.Native;
+
+namespace Jint.Runtime.Debugger
+{
+    /// <summary>
+    /// Dictionary of bindings for a single scope in the scope chain
+    /// </summary>
+    public class DebugScope : IReadOnlyDictionary<string, JsValue>
+    {
+        private IReadOnlyDictionary<string, JsValue> variables;
```

Comment:
could probably have this as `Dictionary<string, JsValue>` as construction is internal, concrete type as field gives best access performance

Reply:
This one is a "bit" unfinished, since we may not want a dictionary in the first place. It's somewhat left over, and probably overkill - the only real reason for its creation is to leave out shadowed bindings in outer scopes. But that might as easily be done with just a List of the names, with value being retrieved on-demand. Which would also be better for the most common use case - iteration of all of the bindings (for display in `Scopes` panel). Especially since we may want to order them in the future. For example, currently `Global` scope lists all intrinsic global properties first, with user variables - which are arguably more "important" - coming last.

Reply:
Thinking further on the API of DebugScope...

I may drop the flattening of `FunctionEnvironmentRecord` and its child `DeclarativeEnvironmentRecord` into "Local/Closure" scope, and let it be up to the caller to flatten them if wanted. Although V8 does this flattening, and it's nice for clarity, I think it's mainly a result of internal V8 optimization rather than a "user friendliness choice". Basically like how Jint includes `Script` top-level block scoped bindings in `Global`. And how V8 also removes closure bindings that aren't needed.

And the flattening adds some amount of unneeded complexity to the iteration of bindings - and inefficiency to random access value retrieval - in each scope:

Since each binding name may be resolved in a different record, at the very least the `DebugScope` would need to iterate over each of its records, then their associated binding names, just for enumeration. And random access retrieval of value would require a `HasBinding` call (the most efficient way I can think of, most often just involving a Dictionary lookup - rather than needing to look up in the ordered list of binding names in `DebugScope`).

So, probably better to have 1 environment record per scope. There may also be cases where the caller would *want* more granular scopes.

The only issue when the caller *wants* to flatten the scopes, is that because `DebugScopes` leaves out empty scopes, you cannot distinguish top-level block scoped variables (which should be flattened) from an empty top-level with block scoped variables in a `BlockStatement` inside. But that could be solved with a boolean `TopLevel` property on the `DebugScope`.

Might also leave the addition of `this` to the Local scope to the caller - since some callers may not actually want it included (devtools doesn't) - and it's available from `CallFrame.This` (where devtools wants it).

Reply:
Clarifying the flattening vs no flattening visually:

Flattened a la Chromium:
![image](https://user-images.githubusercontent.com/1401346/112759540-bb789980-8ff3-11eb-9447-d201d9ccd9c5.png)

Not flattened:
![image](https://user-images.githubusercontent.com/1401346/112759551-cb907900-8ff3-11eb-83cb-1d3b97ca87e2.png)